File: php7/test/GildedRoseTest.php
<?php

namespace App;

require '../vendor/autoload.php';

use PHPUnit\Framework\TestCase;

class GildedRoseTest extends TestCase
{

    public function testSellInDecreasesByOne()
    {
        $items = [new Item("foo", 1, 0)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->sell_in);
    }

    public function testQualityDecreasesByOne()
    {
        $items = [new Item("foo", 0, 1)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->quality);
    }

    public function testQualityNeverNegative()
    {
        $items = [new Item("foo", 0, 1)];
        $gildedRose = new GildedRose($items);
        $days = 5;
        for ($days; $days > 0; $days--) {
            $gildedRose->updateQuality();
        }
        $this->assertEquals(0, $items[0]->quality);
    }

    public function testQualityNeverOverFifty()
    {
        $items = [new Item("foo", 0, 60)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(50, $items[0]->quality);
    }

    public function testSulfurasSellInNeverChanges()
    {
        $items = [new Item("Sulfuras, Hand of Ragnaros", 0, 80)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->sell_in);
    }

    public function testSulfurasQualityNeverChanges()
    {
        $items = [new Item("Sulfuras, Hand of Ragnaros", 0, 80)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(80, $items[0]->quality);
    }

    public function testAgedBrieQualityIncreasesByOne()
    {
        $items = [new Item("Aged Brie", 1, 0)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(1, $items[0]->quality);
    }

    public function testAgedBrieQualityIncreasesByTwoIfSellInIsNegative()
    {
        $items = [new Item("Aged Brie", 0, 0)];
        $gildedRose = new GildedRose($items);
        $days = 5;
        for ($days; $days > 0; $days--) {
            $gildedRose->updateQuality();
        }
        $this->assertEquals(10, $items[0]->quality);
    }

    public function testBackstagePassesQualityIncreasesByTwo()
    {
        $items = [new Item("Backstage passes to a TAFKAL80ETC concert", 10, 0)];
        $gildedRose = new GildedRose($items);
        $days = 10;
        for ($days; $days > 5; $days--) {
            $gildedRose->updateQuality();
        }
        $this->assertEquals(10, $items[0]->quality);
    }

    public function testBackstagePassesQualityIncreasesByThree()
    {
        $items = [new Item("Backstage passes to a TAFKAL80ETC concert", 5, 0)];
        $gildedRose = new GildedRose($items);
        $days = 5;
        for ($days; $days > 0; $days--) {
            $gildedRose->updateQuality();
        }
        $this->assertEquals(15, $items[0]->quality);
    }

    public function testBackstagePassesQualityChangesToZero()
    {
        $items = [new Item("Backstage passes to a TAFKAL80ETC concert", 0, 10)];
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->quality);
    }

    public function testConjuredDecreasesByTwo()
    {
        $items = [new Item("Conjured Mana Cake", 5, 10)];
        $gildedRose = new GildedRose($items);
        $days = 5;
        for ($days; $days > 0; $days--) {
            $gildedRose->updateQuality();
        }
        $this->assertEquals(0, $items[0]->quality);
    }

    public function testConjuredDecreasesByFour()
    {
        $items = [new Item("Conjured Mana Cake", 0, 20)];
        $gildedRose = new GildedRose($items);
        $days = 5;
        for ($days; $days > 0; $days--) {
            $gildedRose->updateQuality();
        }
        $this->assertEquals(0, $items[0]->quality);
    }


}

File: Java/src/test/java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import static org.junit.Assert.*;

import org.junit.Test;

public class GildedRoseTest {

    @Test
    public void testSellInDecreasesByOne() {
        Item[] items = new Item[]{new Item("foo", 1, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].sellIn);
    }

    @Test
    public void testQualityDecreasesByOne() {
        Item[] items = new Item[]{new Item("foo", 1, 1)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void testQualityNeverNegative() {
        Item[] items = new Item[]{new Item("foo", 0, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void testQualityDecreasesByTwoWhenSellInIsNegative() {
        Item[] items = new Item[]{new Item("foo", 0, 2)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void testQualityNeverOverFifty() {
        Item[] items = new Item[]{new Item("Aged Brie", 0, 50)};
        GildedRose app = new GildedRose(items);

        for (int days = 5; days > 0; days--) {
            app.updateQuality();
        }
        assertEquals(50, app.items[0].quality);
    }


    @Test
    public void testSulfurasSellInNeverChanges() {
        Item[] items = new Item[]{new Item("Sulfuras, Hand of Ragnaros", 0, 80)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].sellIn);
    }

    @Test
    public void testSulfurasQualityNeverChanges() {
        Item[] items = new Item[]{new Item("Sulfuras, Hand of Ragnaros", 0, 80)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(80, app.items[0].quality);
    }

    @Test
    public void testAgedBrieQualityIncreasesByOne() {
        Item[] items = new Item[]{new Item("Aged Brie", 1, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(1, app.items[0].quality);
    }

    @Test
    public void testAgedBrieQualityIncreasesByTwoIfSellInIsNegative() {
        Item[] items = new Item[]{new Item("Aged Brie", 0, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(2, app.items[0].quality);
    }

    @Test
    public void testBackstagePassesQualityIncreasesByOne() {
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 12, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(1, app.items[0].quality);
    }

    @Test
    public void testBackstagePassesQualityIncreasesByTwo() {
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 11, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(2, app.items[0].quality);
    }

    @Test
    public void testBackstagePassesQualityIncreasesByThree() {
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 6, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(3, app.items[0].quality);
    }

    @Test
    public void testBackstagePassesQualityDropsToZeroWhenSellInIsNegative() {
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 0, 5)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void testConjuredItemQualityDecreasesByTwo() {
        Item[] items = new Item[]{new Item("Conjured Mana Cake", 2, 2)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void testConjuredItemQualityDecreasesByFour() {
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 0, 4)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals(0, app.items[0].quality);
    }
}
